### PR TITLE
Expose PrestoDriverUri class

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoConnection.java [hunk 1]
```diff
@@ -106,8 +106,9 @@
         requireNonNull(uri, "uri is null");
         this.jdbcUri = uri.getJdbcUri();
         this.httpUri = uri.getHttpUri();
-        this.schema.set(uri.getSchema());
-        this.catalog.set(uri.getCatalog());
+        uri.getSchema().ifPresent(value -> schema.set(value));
+        uri.getCatalog().ifPresent(value -> catalog.set(value));
+        this.catalog.set(uri.getCatalog().orElse(null));
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoConnection.java:111]: "initialize" them also when uri's provided is empty:

```
this.schema.set(uri.getSchema().orElse(null))
...
```

also, catalog is set twice

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java [hunk 2]
```diff
@@ -81,7 +81,7 @@
 /**
  * Parses and extracts parameters from a Presto JDBC URL.
  */
-final class PrestoDriverUri
+public final class PrestoDriverUri
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java:85]: Add rationale to the cmt mesg. Perhaps
```
This allows applications using Presto JDBC to validate connection string without need to
open a connection, or when opening a connection is not possible.
```

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java [hunk 3]
```diff
@@ -343,6 +344,7 @@ private void initCatalogAndSchema()
         path = path.substring(1);
 
         List<String> parts = Splitter.on("/").splitToList(path);
+
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java:347]: unrelated

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java [hunk 4]
```diff
@@ -352,16 +354,22 @@ private void initCatalogAndSchema()
...
+        if (Strings.isNullOrEmpty(parts.get(0))) {
             throw new SQLException("Catalog name is empty: " + uri);
         }
-        catalog = parts.get(0);
+
+        catalog = Optional.ofNullable(parts.get(0));
 
         if (parts.size() > 1) {
-            if (parts.get(1).isEmpty()) {
+            if (Strings.isNullOrEmpty(parts.get(1))) {
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java:368]: unrelated?

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoConnection.java (hunk 1 above)
Comment (wendigo) [reply presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoConnection.java:111]: AtomicReference is null by default

Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoConnection.java:111]: the this.catalog is set twice

**Comment by wendigo** [comment]
AC @findepi

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java [hunk 5]
```diff
@@ -84,6 +84,7 @@
 public final class PrestoDriverUri
 {
     private static final String JDBC_URL_START = "jdbc:";
+    private static final String JDBC_SCHEME = "presto";
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java:87]: We have this validation in PrestoDriver, see `io.prestosql.jdbc.PrestoDriver#DRIVER_URL_START`

Of course, now that `PrestoDriverUri` is public, it needs its own validation.

- make the `Make PrestoDriverUri accessible outside of package` commit last, since all others are actually prerequisites for that
- in this validation here, use `io.prestosql.jdbc.PrestoDriver#DRIVER_URL_START` const (move it from driver to here)

**Review by findepi — CHANGES_REQUESTED** [review]

**Comment by wendigo** [comment]
AC @findepi

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestPrestoDriverUri.java [hunk 6]
```diff
@@ -68,17 +77,17 @@ public void testInvalidUrls()
         assertInvalid("jdbc:presto://localhost:8080/hive/default?SSL=true&SSLVerification=", "Connection property 'SSLVerification' value is empty");
 
         // property in url multiple times
-        assertInvalid("presto://localhost:8080/blackhole?password=a&password=b", "Connection property 'password' is in URL multiple times");
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/TestPrestoDriverUri.java:70]: Please add a test case verifying "presto://..." is not a valid URL
(apparently before your change it was parsed successfully)

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriver.java [hunk 7]
```diff
@@ -93,7 +91,7 @@ public Connection connect(String url, Properties info)
     public boolean acceptsURL(String url)
             throws SQLException
     {
-        return url.startsWith(DRIVER_URL_START);
+        return PrestoDriverUri.acceptsURL(url);
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriver.java:94]: from cmt msg:

> Previously we would parse uri with invalid scheme and without jdbc: prefix
that could cause java.lang.StringIndexOutOfBoundsException

this is not quite accurate, since the code was not reachable

also, the cmt title could be improved to suggest the validation is _moved_ not _added:

```
Verify if Presto JDBC url is not malformed within PrestoDriverUri
```

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java [hunk 8]
```diff
@@ -119,6 +119,18 @@ private PrestoDriverUri(URI uri, Properties driverProperties)
         initCatalogAndSchema();
     }
 
+    public static PrestoDriverUri create(String url, Properties properties)
+            throws SQLException
+    {
+        return new PrestoDriverUri(url, properties);
+    }
+
+    public static PrestoDriverUri create(String url)
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoDriverUri.java:128]: We do not need this overload, do we?

**Review by findepi — APPROVED** [review]

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
CI https://github.com/prestosql/presto/issues/5758#issuecomment-745106976
